platform/android/java/lib/src/org/godotengine/godot/utils/PermissionsUtil.java

```diff
@@ -0,0 +1,118 @@
+	 */
+	public static boolean requestPermission(String p_name, Godot p_activity) {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+			// Not necessary, asked on install already
+			return true;
+		}
+
+		if (p_name.equals("RECORD_AUDIO")) {
+			if (ContextCompat.checkSelfPermission(p_activity, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
+				p_activity.requestPermissions(new String[] { Manifest.permission.RECORD_AUDIO }, REQUEST_RECORD_AUDIO_PERMISSION);
+				return false;
+			}
+		}
+
+		if (p_name.equals("CAMERA")) {
+			if (ContextCompat.checkSelfPermission(p_activity, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
+				p_activity.requestPermissions(new String[] { Manifest.permission.CAMERA }, REQUEST_CAMERA_PERMISSION);
+				return false;
+			}
+		}
+
+		if (p_name.equals("VIBRATE")) {
+			if (ContextCompat.checkSelfPermission(p_activity, Manifest.permission.VIBRATE) != PackageManager.PERMISSION_GRANTED) {
+				p_activity.requestPermissions(new String[] { Manifest.permission.VIBRATE }, REQUEST_VIBRATE_PERMISSION);
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Request dangerous permissions which have defined in AndroidManifest.xml. If some of selected
+	 * permissions protection level is <b>dangerous</b>, with this method you can request permissions
+	 * from user.
+	 * @param p_activity the caller activity name for this method.
+	 * @return true/false. "true" if all of permissions were granted otherwise returns "false".
+	 */
+	public static boolean requestManifestPermissions(Godot p_activity) {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+			return true;
+		}
+
+		PackageManager packageManager;
+		PackageInfo packageInfo;
+		try {
+			packageManager = p_activity.getPackageManager();
+			packageInfo = packageManager.getPackageInfo(p_activity.getPackageName(),
+					PackageManager.GET_PERMISSIONS);
+		} catch (PackageManager.NameNotFoundException e) {
+			e.printStackTrace();
+			return false;
+		}
+
+		String[] manifestPermissions = packageInfo.requestedPermissions;
+		if (manifestPermissions == null || manifestPermissions.length == 0)
+			return true;
+
+		List<String> dangerousPermissions = new ArrayList<>();
+		for (String manifestPermission : manifestPermissions) {
+			try {
+				if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+					if (packageManager.getPermissionInfo(manifestPermission, 0).getProtection() == PermissionInfo.PROTECTION_DANGEROUS && ContextCompat.checkSelfPermission(p_activity, manifestPermission) != PackageManager.PERMISSION_GRANTED) {
+						dangerousPermissions.add(manifestPermission);
+					}
+				} else {
+					if (packageManager.getPermissionInfo(manifestPermission, 0).protectionLevel == PermissionInfo.PROTECTION_DANGEROUS && ContextCompat.checkSelfPermission(p_activity, manifestPermission) != PackageManager.PERMISSION_GRANTED) {
+						dangerousPermissions.add(manifestPermission);
+					}
+				}
```

Comment:
You can simplify this logic as follow:
```
PermissionInfo permissionInfo = packageManager.getPermissionInfo(manifestPermission, 0);
int protectionLevel = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P ? permissionInfo.getProtection() : permissionInfo.protectionLevel;
if (protectionLevel == PermissionInfo.PROTECTION_DANGEROUS && 
    ContextCompat.checkSelfPermission(p_activity, manifestPermission) != PackageManager.PERMISSION_GRANTED) {
        dangerousPermissions.add(manifestPermission);
}
```

Reply:
I think your and my logic has same complexity level but I edited according to yours and merge some nested ifs. I analyze code with Sonarlint.

Reply:
Yes the complexity level is the same. The suggestion however allows to reduce the amount of duplicated code.
For example, let's say we'd like to add an additional check or more logic before adding the permission to the `dangerousPermissions` list, with the original logic, the added code would need to be duplicated twice. And this could cause subtle bugs where some code is added in one branch and forgotten to be added in the other.

Reply:
Yes, It's more clear :) Thank you.